File: backend/platform-core/src/main/java/com/commonsengine/platform/support/Enums.java
package com.commonsengine.platform.support;

import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Enum parsing utilities.
 */
public final class Enums {

    private Enums() {
        throw new AssertionError("Utility class — do not instantiate");
    }

    /**
     * Parse a single enum constant, throwing with a helpful list of valid
     * candidates on invalid input.
     */
    public static <T extends Enum<T>> T parse(Class<T> enumType, String value) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(
                    "Cannot parse blank value into enum " + enumType.getSimpleName());
        }
        try {
            return Enum.valueOf(enumType, value.trim().toUpperCase());
        } catch (IllegalArgumentException ex) {
            String candidates = java.util.Arrays.stream(enumType.getEnumConstants())
                    .map(Enum::name)
                    .collect(java.util.stream.Collectors.joining(", "));
            throw new IllegalArgumentException(
                    "Invalid " + enumType.getSimpleName() + " value: '" + value
                            + "'. Valid values: " + candidates);
        }
    }

    /**
     * Parse a collection of enum constant names into a {@link Set}, preserving
     * insertion order. Duplicate or invalid entries throw.
     */
    public static <T extends Enum<T>> Set<T> parseAll(Class<T> enumType, Collection<String> values) {
        if (values == null) {
            return Set.of();
        }
        Set<T> result = new LinkedHashSet<>();
        for (String v : values) {
            result.add(parse(enumType, v));
        }
        return Set.copyOf(result);
    }
}

File: backend/platform-core/src/main/kotlin/com/commonsengine/platform/support/Enums.kt
package com.commonsengine.platform.support

/**
 * 统一的枚举解析工具（#67）。
 *
 * 项目中所有 Controller 的字符串→枚举转换都应通过本工具完成，
 * 保证「非法枚举值 → 400 Bad Request」的一致行为：
 *
 * - 不安全：`EnumType.valueOf(s)` — 非法值抛 `IllegalArgumentException`，
 *   被 `GlobalExceptionHandler` 捕获为 400 ✅，但错误信息不友好（仅含枚举名）。
 * - 静默丢弃：`runCatching { valueOf(s) }.getOrNull()` — 非法值被忽略 ❌
 * - 静默默认：`runCatching { valueOf(s) }.getOrDefault(DEFAULT)` — 静默改写 ❌
 *
 * 本工具统一为：非法值抛出**带候选值清单**的 `IllegalArgumentException`，
 * 由全局异常处理器映射为 400。
 */
object Enums {

    /**
     * 解析枚举。非法值抛 [IllegalArgumentException]（→ 400）。
     *
     * 用法：`Enums.parse<RatingDirection>(body.direction)`
     */
    inline fun <reified T : Enum<T>> parse(value: String): T {
        return runCatching { enumValueOf<T>(value) }.getOrElse {
            throw IllegalArgumentException(
                "无效的 ${T::class.simpleName} 值: '$value'。" +
                    "合法值: ${enumValues<T>().joinToString(", ")}"
            )
        }
    }

    /**
     * 批量解析枚举列表。任一非法值即抛 [IllegalArgumentException]（→ 400）。
     *
     * 用法：`Enums.parseAll<RatingTag>(body.tags)`
     */
    inline fun <reified T : Enum<T>> parseAll(values: Collection<String>): Set<T> =
        values.map { parse<T>(it) }.toSet()
}
